classes/imx-boot-container.bbclass

```diff
@@ -0,0 +1,106 @@
+ATF_MACHINE_NAME = "bl31-${ATF_PLATFORM}.bin"
+ATF_MACHINE_NAME_append = "${@bb.utils.contains('MACHINE_FEATURES', 'optee', '-optee', '', d)}"
+
+# This package aggregates output deployed by other packages, so set the
+# appropriate dependencies
+do_resolve_and_populate_binaries[depends] += " \
+    ${@' '.join('%s:do_deploy' % r for r in '${IMX_EXTRA_FIRMWARE}'.split() )} \
+    imx-atf:do_deploy \
+    ${@bb.utils.contains('MACHINE_FEATURES', 'optee', 'optee-os:do_deploy', '', d)} \
+"
+
+# Append make flags to include ATF load address
+EXTRA_OEMAKE += "ATF_LOAD_ADDR=${ATF_LOAD_ADDR}"
+
+# Define an additional task that collects binary output from dependent packages
+# and depolys them into the U-Boot build folder
+do_resolve_and_populate_binaries() {
+    if [ ! -n "${ATF_LOAD_ADDR}" ]; then
+        bberror "ATF_LOAD_ADDR is undefined, result binary would be unusable!"
+    fi
+
+    if [ -n "${UBOOT_CONFIG}" ]; then
+        for config in ${UBOOT_MACHINE}; do
+            i=$(expr $i + 1);
+            for type in ${UBOOT_CONFIG}; do
+                j=$(expr $j + 1);
+                if [ $j -eq $i ]; then
+                    for ddr_firmware in ${DDR_FIRMWARE_NAME}; do
+                        # Sanitize the FW name as U-Boot expects it to be without version
+                        if [ -n "${DDR_FIRMWARE_VERSION}" ]; then
+                            ddr_firmware_name=$(echo $ddr_firmware | sed s/_${DDR_FIRMWARE_VERSION}//)
+                        else
+                            ddr_firmware_name="$ddr_firmware"
+                        fi
+                        bbnote "Copy ddr_firmware: ${ddr_firmware} from ${DEPLOY_DIR_IMAGE} -> ${B}/${config}/${ddr_firmware_name}"
+                        cp ${DEPLOY_DIR_IMAGE}/${ddr_firmware} ${B}/${config}/${ddr_firmware_name}
+                    done
+                    if [ -n "${ATF_MACHINE_NAME}" ]; then
+                        cp ${DEPLOY_DIR_IMAGE}/${BOOT_TOOLS}/${ATF_MACHINE_NAME} ${B}/${config}/bl31.bin
+                    else
+                        bberror "ATF binary is undefined, result binary would be unusable!"
+                    fi
+                fi
+            done
+            unset  j
+        done
+        unset  i
+    fi
+}
+addtask do_resolve_and_populate_binaries before do_compile after do_configure
+
+# Append the u-boot do_deploy task to deploy also the result flash.bin
+# boot container
+do_deploy_append() {
+    # Deploy the resulted flash.bin for WIC to pick it up
+    if [ -n "${UBOOT_CONFIG}" ]; then
+        for config in ${UBOOT_MACHINE}; do
+            i=$(expr $i + 1);
+            for type in ${UBOOT_CONFIG}; do
+                j=$(expr $j + 1);
+                if [ $j -eq $i ]
+                then
+                    install -m 0777 ${B}/${config}/flash.bin  ${DEPLOYDIR}/flash.bin-${MACHINE}-${UBOOT_CONFIG}
+                    ln -sf flash.bin-${MACHINE}-${UBOOT_CONFIG} flash.bin
```

Comment:
I believe the `flash.bin` here is closer to what U-Boot mainline uses. I have mixed opinions about this.

Reply:
@MaxKrummenacher This could potentially be done, thus reducing the implementation.

My original idea however was to completely separate `imx-boot` from `flash.bin` all the way down to WKS file. By this, we would have a independent implementation for boot container, and in case NXP would change their WKS generation - we would not be dependent on it.

Second, renaming `flash.bin` to `imx-boot` in `do_deploy` task would rather obscure the Community part of boot container, thus making it not so visible for those who utilizes it. It is not a major issue, rather a clean readability.

I tend to agree with @otavio here that the path from U-Boot providing `flash.bin` should rather be kept here.

If there is a strong objection to those points above - I can address it further in _V2_ though.

Reply:
Well, flash.bin is also  the file name created by imx-boot/imx-mkimage. It only gets renamed in the OE recipe.
And the boot containers from a build directly from U-Boot vs. the one where the assembly is done with imx-boot/imx-mkimage should be identical.
Thus I think for what comes after the creation of the boot container having one name simplifies matters and does not complicate it. E.g. one wks file instead of two, one uuu script instead of two ...